packages/eslint-plugin/tests/rules/switch-exhaustiveness-check.test.ts

```diff
@@ -0,0 +1,292 @@
+
+const ruleTester = new RuleTester({
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    tsconfigRootDir: rootPath,
+    project: './tsconfig.json',
+  },
+});
+
+ruleTester.run('switch-exhaustiveness-check', switchExhaustivenessCheck, {
+  valid: [
+    // All branches matched
+    `
+type Day = 'Monday' | 'Tuesday' | 'Wednesday' | 'Thursday' | 'Friday' | 'Saturday' | 'Sunday'
+
+const day = 'Monday' as Day
+let result = 0
+
+switch (day) {
+  case 'Monday': {
+     result = 1
+     break
+  }
+  case 'Tuesday': {
+     result = 2
+     break
+  }
+  case 'Wednesday': {
+     result = 3
+     break
+  }
+  case 'Thursday': {
+     result = 4
+     break
+  }
+  case 'Friday': {
+     result = 5
+     break
+  }
+  case 'Saturday': {
+     result = 6
+     break
+  }
+  case 'Sunday': {
+     result = 7
+     break
+  }
+} 
+`,
+    // Other primitive literals work too
+    `
+type Num = 0 | 1 | 2
+type Bool = true | false
+type Mix = 0 | 1 | 'two' | 'three' | true
+
+function test1(value: Num): number {
+  switch (value) {
+    case 0: return 0
+    case 1: return 1
+    case 2: return 2
+  }
+}
+
+function test2(value: Bool): number {
+  switch (value) {
+    case true: return 1
+    case false: return 0
+  }
+}
+
+function test3(value: Mix): number {
+  switch (value) {
+    case 0: return 0
+    case 1: return 1
+    case 'two': return 2
+    case 'three': return 3
+    case true: return 4
+  }
+}
+`,
+    // Works with type references
+    `
+type A = 'a'
+type B = 'b'
+type C = 'c'
+type Union = A | B | C
+
+function test(value: Union): number {
+  switch (value) {
+    case 'a': return 1
+    case 'b': return 2
+    case 'c': return 3
+  }
+}
+`,
+    // Works with `typeof`
+    `
+const A = 'a'
+const B = 1
+const C = true
+
+type Union = typeof A | typeof B | typeof C
+
+function test(value: Union): number {
+  switch (value) {
+    case 'a': return 1
+    case 1: return 2
+    case true: return 3
+  }
+}
+`,
+    // Switch contains default clause.
+    `
+type Day = 'Monday' | 'Tuesday' | 'Wednesday' | 'Thursday' | 'Friday' | 'Saturday' | 'Sunday'  
+
+const day = 'Monday' as Day
+let result = 0
+
+switch (day) {
+  case 'Monday': {
+     result = 1
+     break
+  }
+  default: {
+     result = 42
+  }   
+}
+  `,
+    // Exhaustiveness check only works for union types...
+    `
+const day = 'Monday' as string
+let result = 0
+
+switch (day) {
+  case 'Monday': {
+    result = 1
+    break
+  }
+  case 'Tuesday': {
+    result = 2
+    break
+  }
+} 
+  `,
+    // ... and enums (at least for now).
+    `
+enum Enum { A, B }
+
+function test(value: Enum): number {
+  switch (value) {
+    case Enum.A: return 1
+    case Enum.B: return 2
+  }
+}
+`,
+    // Object union types won't work either, unless it's a discriminated union
+    `
+type ObjectUnion = { a: 1 } | { b: 2 }
+
+function test(value: ObjectUnion): number {
+  switch (value.a) {
+    case 1: return 1
+  }
+}
+`,
+  ],
+  invalid: [
```

Comment:
Because the logic which creates it is "non-trivial", could please you test the `data` of each error?

```ts
{
  messageId: 'switchIsNotExhaustive',
  data: {
    missingBranches: "'Tuesday' | 'Wednesday' | 'Thursday' | 'Friday' | 'Saturday' | 'Sunday'",
  },
  ...
}
```